File: Difficulty/Medium/[200] Number of Islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid: return 0;
        nr, nc, num = len(grid), len(grid[0]), 0
        for i in range(nr):
            for j in range(nc):
                if grid[i][j] != '0':
                    num += 1
                    self.dfs(grid, i, j)
        return num

    def dfs(self, grid, x, y):
        nr, nc = len(grid), len(grid[0])
        grid[x][y] = '0'
        if x - 1 >= 0 and grid[x-1][y] == '1':
            self.dfs(grid, x-1, y)
        if x + 1 < nr and grid[x+1][y] == '1':
            self.dfs(grid, x+1, y)
        if y - 1 >= 0 and grid[x][y-1] == '1':
            self.dfs(grid, x, y-1)
        if y + 1 < nc and grid[x][y+1] == '1':
            self.dfs(grid, x, y+1)

File: Difficulty/Medium/[200] Number of Islands/solution.cpp
/*
 * @lc app=leetcode id=200 lang=cpp
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        unordered_set<string> seen;
        int res = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] != '0') {
                    if (seen.find(to_string(i) + '_' + to_string(j)) == seen.end()) {
                        dfs(grid, seen, i, j);
                        res++;
                    }
                }   
            }
        }
        return res;
    }

    void dfs(vector<vector<char>>& grid, unordered_set<string>& seen, int x, int y) {
        if (x < 0 || y < 0 || x >= grid.size() || y >= grid[0].size()) return;
        if (grid[x][y] != '0') {
            string cord = to_string(x) + '_' + to_string(y);
            if (seen.find(cord) == seen.end()) {
                seen.insert(cord);
                dfs(grid, seen, x + 1, y);
                dfs(grid, seen, x - 1, y);
                dfs(grid, seen, x, y + 1);
                dfs(grid, seen, x, y - 1);
            }
        }
    }
};
// @lc code=end

